aten/src/ATen/VmapTransforms.cpp

```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
   return bdims;
 }
 
-static bool batchDimsAreSameSize(
-    const Tensor& physical,
-    const Tensor& reference,
-    int64_t num_batch_dims) {
-  auto physical_sizes = physical.sizes();
-  auto reference_sizes = reference.sizes();
-  if (physical_sizes.size() < num_batch_dims) {
-    return false;
+Tensor VmapPhysicalView::newLogicalFromPhysical(const Tensor& physical) {
+  return makeBatched(physical, computeFrontBatchDimsFromLevels(levels_));
+}
+
+// Given a Tensor or a BatchedTensor, returns the underlying physical tensor
+// with all vmapped dimensions permuted to the front, if they exist, and a
+// bitset of vmap levels that were present in the tensor.
+static std::pair<Tensor,std::bitset<kVmapNumLevels>>
+getPhysicalTensorAndLevels(const Tensor& self) {
+  auto* batched = maybeGetBatched(self);
+  if (batched) {
+    return {permuteBatchDimsToFront(batched), createLevelsBitset(batched->bdims())};
   }
-  if (reference_sizes.size() < num_batch_dims) {
-    return false;
+  return {self, 0};
+}
+
+// Given a Tensor or a BatchedTensor, creates a physical view of the tensor
+// such that it has a batch dimension for each levels in `requested_levels`
+// and `requested_example_dim` number of non-batch-dimensions.
```

Comment:
It will be easier to understand if we explain why you want to do this. An example would help too.